packages/utils/src/arrayToDictionary.ts

```diff
@@ -1,32 +1,50 @@
+type DictionaryKey = string | number;
+type GetKey<T> = (item: T) => any;
+type Key<Fn extends (...a: any[]) => any, R = ReturnType<Fn>> = R extends DictionaryKey ? R : never;
 type ArrayToDictionary = {
-    <T>(array: T[], getKey: (item: T) => string | number, multiple?: false): { [key: string]: T };
-    <T>(array: T[], getKey: (item: T) => string | number, multiple: true): { [key: string]: T[] };
+    <T, Fn extends GetKey<T>>(array: T[], getKey: Fn, multiple?: false): Record<Key<Fn>, T>;
+    <T, Fn extends GetKey<T>>(array: T[], getKey: Fn, multiple: true): Record<Key<Fn>, T[]>;
 };
 
 /**
  * @param array Array to be converted to dictionary
  * @param getKey Function extracting string from an array item T, which will become its
  * key in the dictionary (if not unique, latter item could replace the former one)
+ * Item will not be added to dictionary if key is not defined
  * @param multiple If true, dictionary values are arrays of all items with the given key
  * @returns Dictionary object with array items as values
  */
-export const arrayToDictionary: ArrayToDictionary = <T>(
+
+const validateKey = (key: any): key is DictionaryKey => {
```

Comment:
nit: I think we should start using `unknown` more, for situations where a function actually _doesn't know_ the incoming type and not when it _doesn't care_ about the type. `validateKey()` does care about the type eventually since it does the type check and returns based on the outcome. Overall, `any` is not even a type but a way to bypass type-checking altogether as I see it :D With better type safety, `unknown` could help to catch possible errors early on. What do you say? 

Btw here it obviously would not affect anything, more of a stylistic choice but getting used to `unknown` in general could still be useful.

Reply:
i dont have a strong opinion on that, it that particular case it doesnt really matter.
changed to uknown, rebased and squashed

Reply:
Perfect! Added to the Code Style's [proposals section](https://www.notion.so/satoshilabs/Proposals-6296b830b8e94df889ba93f66f7964a5).